test/RachisTests/DatabaseCluster/AtomicClusterReadWriteTests.cs

```diff
@@ -0,0 +1,281 @@
+                await session.SaveChangesAsync();
+            }
+            await WaitForDocumentInClusterAsync<TestObj>(documentStore.GetRequestExecutor().Topology.Nodes, entity.Id, u => u != null, TimeSpan.FromSeconds(10));
+
+            var barrier = new Barrier(3);
+            var exceptions = new ConcurrentBag<Exception>();
+            var tasks = Enumerable.Range(0, stores.Length)
+                .Select(i => Task.Run(async () =>
+                {
+                    using var session = stores[i].OpenAsyncSession(new SessionOptions {TransactionMode = TransactionMode.ClusterWide});
+                    var loaded = await session.LoadAsync<TestObj>(entity.Id);
+                    barrier.SignalAndWait();
+
+                    loaded.Prop = "Change" + i;
+
+                    try
+                    {
+                        await session.SaveChangesAsync();
+                    }
+                    catch (Exception e)
+                    {
+                        exceptions.Add(e);
+                    }
+                }));
+
+            await Task.WhenAll(tasks);
+            Assert.Equal(2, exceptions.Count);
+            foreach (var exception in exceptions)
+            {
+                Assert.IsType<ConcurrencyException>(exception);
+            }
+        }
+
+        [Fact]
+        public async Task ClusterWideTransaction_WhenLoadAndDeleteWhileUpdated_ShouldSucceedOnlyInTheFirst()
+        {
+            var (nodes, leader) = await CreateRaftCluster(3, shouldRunInMemory: false);
+
+            using var documentStore = GetDocumentStore(new Options {Server = leader, ReplicationFactor = nodes.Count});
+
+            var entity = new TestObj();
+            using (var session = documentStore.OpenAsyncSession(new SessionOptions { TransactionMode = TransactionMode.ClusterWide }))
+            {
+                await session.StoreAsync(entity);
+                await session.SaveChangesAsync();
+            }
+
+            await LoadAndDeleteWhileUpdated(nodes, documentStore.Database, entity.Id);
+        }
+
+        [Fact]
+        public async Task ClusterWideTransaction_WhenImportThenLoadAndDeleteWhileUpdated_ShouldSucceedOnlyInTheFirst()
+        {
+            var (nodes, leader) = await CreateRaftCluster(3);
+            using var documentStore = GetDocumentStore(new Options { Server = leader, ReplicationFactor = nodes.Count });
+
+            var entity = new TestObj();
+            using (var source = GetDocumentStore())
+            {
+                using var session = source.OpenAsyncSession(new SessionOptions { TransactionMode = TransactionMode.ClusterWide });
+                await session.StoreAsync(entity);
+                await session.SaveChangesAsync();
+
+                var operation = await source.Smuggler.ExportAsync(new DatabaseSmugglerExportOptions(), documentStore.Smuggler);
+                await operation.WaitForCompletionAsync(TimeSpan.FromMinutes(1));
```

Comment:
What is this for?

Reply:
I tried to think about edge cases.
An external source for a cluster-wide transaction document seemed to me as one?